Create CHANGELOG.md and VERSIONING.md document (#2284)
chipsalliance/rocket-chip: VERSIONING.md

= Rocket Chip Versioning Strategy

== Version Naming

* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).

> * The format is YYYY.MM.DD.MINOR

Reviewer: Just to be clear, the major release designation is the `YYYY.MM.DD` component. Minor releases will retain this designation (regardless of their release date), and simply bump the `MINOR` component. Perhaps a couple of examples would help clarify this.

Author: ah yes good point

Reviewer 2: I just want to mention that this violates two different parts of CalVer, so we probably shouldn't call this CalVer:

- Major, Minor, and Micro are used to describe the ordinal position of the version component, with Minor specifically being used to refer to the second position.
- They discourage the use of 4-component versions, and there is nothing that matches the component that you are indicating with the "MINOR" designation above.

Author: "Major, Minor, and Micro are used to describe the ordinal position of the version component, with Minor specifically being used to refer to the second position.
"
Hm, that didn't match my read of the calver link posted above, which called out several examples of projects using various combinations of these things, but perhaps i should read it more closely to see if they still considered those "calver" or "calver-inspired"

Author: Ok, I see what you are saying. The only example on their site which violates the "dot-position of MAJOR/MINOR/MICRO defines what it is" is Teradata, which they explicitly call out a having a "useful combination of semvar and calver":
<img width="239" alt="Screen Shot 2020-04-04 at 2 59 32 AM" src="https://user-images.githubusercontent.com/17858596/78424176-6abca380-7620-11ea-9090-d6854163b7e5.png">

Author: There is also this one that seems to use an `_` to get around the "MICRO is in the wrong spot" but they don't give any further explanation on whether that is encouraged/discouraged:
<img width="856" alt="Screen Shot 2020-04-04 at 3 03 47 AM" src="https://user-images.githubusercontent.com/17858596/78424269-2aa9f080-7621-11ea-9a7f-b31be73f143d.png">

Author: @richardxia do you think that we "shouldn't call this calver" or should change it until it matches calver (which presumably is a good thing)? For example I can't tell if using something like `YYYY_0M_0D.MINOR` would address your points since there aren't a lot of "underscore" examples.

Reviewer 2: I think we should pop up a level to discuss what kinds of compatibility guarantees we want to provide and what our branching strategy and bugfix policy will be. Despite using dates in the format string, you seem to want to provide some notion of semantic versioning by adopting a definition of "minor" that is equivalent to SemVer's. If this is important to us, then we shouldn't obfuscate it using CalVer.

You can see that in the examples on the calver.org website, besides the `MICRO` version components, the projects their either fall into the category of "every non-micro release is guaranteed to break compatibility and we are not trying to address that at all" (e.g. Ubuntu) or "my library is basically permanently backwards compatible, but I am dependent on stuff that changes over time, like time zones or SSL certificate bundles". Most of the libraries on that page which have a `MICRO` version only really use it for bugfixes or security patches.

Note that you _are_ allowed to use SemVer while trying to stick to a date-based release schedule; you can fully adopt SemVer's versioning schemes, and instead allowing changes to dictate SemVer version bumps, you can let your release schedule dictate what kinds of changes are allowed to be made.

Rocket Chip Versioning Strategy ** All Strawman to be discussed**
===============================

Version Naming
--------------

* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).

> * The format is YYYY.MM.DD.MINOR

Author: ```suggestion
* The format is YYYY.MM.DD.MINOR
  * For example, 2020.04.04.0 is the release which occurred on April 4, 2020.
  * 2020.04.04.1 is a release that could occur at any time, but introduces only MINOR changes over the 2020.04.04 release
  * 2020.06.01.0 is a release that occurs on June 1, 2020, with major API breaking changes over the 2020.04.04.X release.
```

Reviewer 2: This does not match the CalVer formatting conventions. Specifically:

> - MM - Short month - 1, 2 ... 11, 12
> - 0M - Zero-padded month - 01, 02 ... 11, 12

You should either update the string to say `YYYY.0M.0D` or update the examples to say `2020.4.4`.

* MINOR changes include:
  * API Additions

>   * Bug Fixes (which can result in changes in the emitted hardware considered buges)

Author: ```suggestion
  * Bug Fixes (which can result in changes in the emitted hardware considered bugs)
```

* MINOR changes do NOT include:
  * Changes in top-level I/Os for < level of hierarchy >
  * Backwards incompatible changes in the API (see below)

API
---

The Rocket-Chip API is: 
  * All public Scala methods and classes

>   * All public Scala methods and classes of included Scala libraries (hardfloat, CDE, etc)

Reviewer 3: It isn't clear to me if the definition of the Scala API implies either source or binary compatibility for minor version bumps (the discussion below about `public val`s in `Module`s seems to imply neither, which is a pragmatic choice but differs from how most scala projects work). I think it would be good to be specific as to whether one of (or any of) binary and source compatibility are intended to be preserved.

I see that mima is in plugins.sbt- is the intention for CI to report binary compatibility for PRs? That would be awesome, even if it can't always be preserved.

Author: > It isn't clear to me if the definition of the Scala API implies either source or binary compatibility for minor version bumps (the discussion below about public vals in Modules seems to imply neither, which is a pragmatic choice but differs from how most scala projects work).

Yeah, this is a big question in my mind. On the one hand, just using "normal" Scala API makes a lot of sense, since in actuality every `val foo = Wire(...)` you create really is an API point for someone else even in scala-land (there are utilities to leverage that to generate cross-module references, etc).

On the other hand, the thing that often causes/cures pain for Chisel consumers is things like:
`val foo = Reg(...).suggestName("a_better_name")`
because it gives an API to the *output* of the generated code.
"a_better_name"  becomes a pretty important part of the API too, but will be harder to detect with binary compatibility.

I'll be honest that I am just getting familiarized with this topic myself, and would be interested in learning more how this binary compatability stuff should be done.

  * The command line used to invoke the generator

>   * All Verilog interfaces (I/Os and parameterization) for verilog files

Reviewer 2: Is this just of the top-level module or all modules?

Author: There is no "top-level module" in rocket-chip which is a library of module generators. So, really I'd have to say "all modules". But yeah not sure how this could be enforced especially with diplomatic punching of things other than the previous requirement that all classes (and `Bundle`s are classes) stay the same, so maybe this is captured in that.

Actually most `val io = IO ( new Bundle {...})` are actually anonymous classes. Not sure how an anonymous class is considered in a "normal" scala API. Presumably if I added/removed a var or changed its type inside that `Bundle` that would be considered a change in the type of `val io` so would be considered a breaking change by my above rules?

In thinking about this more it's probably best to just use Scala standard definition of the API as it will probably capture *most* of the hardware use cases we think about.

Reviewer 2: > There is no "top-level module" in rocket-chip which is a library of module generators.

Let me reword my question, since this is a bit more subtle than I had realized.

For a given "top" module (the actual module that the Chisel generation flow starts with or perhaps also any markDUT module), are the Verilog interfaces of the Verilog modules below it in that particular generation allowed to change?

For example, suppose I had the two Chisel modules (and please excuse the syntax I am making up because I don't actually know Chisel):

```scala
class OuterModule extends Module {
  val io = Bundle { val out = Output(Bool), val in1 = Input(Bool), val in2 = Input(Bool) }
  val inner_module = InnerModule()
  inner_module.io.in1 := io.in1
  inner_module.io.in2 := io.in2
  io.out := inner_module.io.out
}

class InnerModule extends Module {
  val io = Bundle { val out = Output(Bool), val in1 = Input(Bool), val in2 = Input(Bool) }
  io.out := io.in1 && io.in2
}
```

Now, suppose in a future version, we have the following change to `OuterModule`:

```diff
-  inner_module.io.in2 := io.in2
+  inner_module.io.in2 := false.asBool
```

Due to constant propagation, `InnerModule`s interface will likely change due to optimizing away the constant signal.

Now, if you instead chose `InnerModule` as your top-level module for a particular run of the generators, the IOs would _not_ change (I think?).

Author: I realized that this comment was in particular on a line i meant to target the verilog resources in this file (black boxes), not generated verilog.

Author: I am editing the text to say "verilog resources", which may not really resolve this conversation but my intention was the files which we are explicitly including as verilog here.

  * C code...?

>   * The version(s) of Chisel, FIRRTL, rocket-tools pointed to by rocket-chip.

Reviewer 3: Is the intention here to have an `=` or a `>=` relationship with the Chisel/FIRRTL/etc. versions?

Author: i think my intention is if you bump the major version of CHisel/FIRRTL, that would count as a major (not MINOR) change. But bumping the Chisel/FIRRTL version to one that is also backwards compatible would be considered a MINOR change.

Does that answer your question? How can I make it more clear?

Reviewer 3: Ah, I see. That makes sense. I think for sbt this means an `=` relationship, i.e. a newer version of the dep does not satisfy the requirement. It also makes sense that major releases should happen together.

My main reason for asking about this is that sometimes different libraries could want different deps, for example my radio baseband might want a newer version of chisel than rocket, but I want to use them together. Because of the way binary compatibility works with chisel stuff, it would be hard to do the normal scala thing and have a `>=` dependency constraint. It's probably better to explicitly say you are not using a standard release (which may come with a recommendation that you re-run tests) if you are bumping deps.

Perhaps it should have subsections that say something like:

* A new minor release of rocket may update firrtl, chisel, and rocket-tools to a newer minor release of those dependencies. 
* A new major release of rocket may update firrtl, chisel, and rocket tools to a newer major release of those dependencies.

Author: @grebe take a look at https://github.com/chipsalliance/rocket-chip/pull/2284/files#diff-baa362176b63e6912fc6721235a68502R41 and let me know if this is more clear to you

  * Any code that generates IO(s) of any Module
  * The cmd line interface to scripts like vlsi-mem-gen

>   * `public val`s within `LazyModule`s which translate to diplomatic nodes (?? -- this could go either way)

Reviewer 2: How about the Makefiles/Makefrags?

Also, what about the Wake files and wit-manifest.json?

Author: > How about the Makefiles/Makefrags?
> Also, what about the Wake files

Good point. Do you have a suggestion? How is this dealt with in other projects?

> and wit-manifest.json?

To me this is like asking "what about the submodules" which I tried to capture as well. Is there something special about wit-manifest.json that should be captured differently than submodules? In particular, shoudl we be explicit about changes to the version of wit/wake needed to deal with this repo (for example, if we make a change to this repo that requires you to bump to a new major version of wit/wake, in my mind that should be a MAJOR change).

Not sure the right classification if we make a change that requires you to bump to a new minor version of wit/wake. I think this is sort of what @grebe was discussing above RE chisel/firrtl, but it's a little inverted because rocket-chip does not *provide* wit/wake wheras rocket-chip is often used as the *provider* of chisel/firrtl to projects which use rocket-chip as a library.

Reviewer 2: > > How about the Makefiles/Makefrags?
> > Also, what about the Wake files
> 
> Good point. Do you have a suggestion? How is this dealt with in other projects?

In my experience, most other software projects do not attempt to rely on the build scripts of their dependencies at all, and they are not part of the "interface". However, I was under the impression that some UCB folks or other open source users may be relying on them.

> Is there something special about wit-manifest.json that should be captured differently than submodules / the repos it points to?

I agree that it's mostly the same as the Git submodules. However, one big technical difference is that the `wit-manifest.json` describes a `>=` relationship, not an `==` relationship, so if in the `wit-manifest.json` you say that you depend on version A of a library, you are really saying that you depend on A or any version that has A as an ancestor. This does make it a little bit easier on versioning since users are allowed to use any version of the dependency that is newer.

Reviewer 4: FWIW, we (chipyard & firesim) do not use Rocket Chip's checked-in makefrags but do use the generated ones.

Author: I added Makefiles/Makefrags to the "NOT part of the API" section, and then tried to clarify the wit-manifest dependencies behavior. Can you take a look @richardxia to see if what I wrote in https://github.com/chipsalliance/rocket-chip/pull/2284/files#diff-baa362176b63e6912fc6721235a68502R47
helps answer your question?

Reviewer 2: That answers my question, but it sounds like chipyard and firesim use the generated Makefiles.

FWIW, SiFive does not use any of rocket-chip Makefiles/Makefrags/generated Makefiles internally, either.

The following is NOT part of the API:
  * `public val`s within `Module`s which translate to Wires, Regs, module instances
    * (We note that this may cause issues for people using Cross-Module references)
  * Internals of scripts like vlsi-mem-gen, the contents of such scripts
  * Internals of any verilog black boxes
  * C Code ...?
      
Roadmap
-------

* Upcoming Versions:

  - Version ~2020.04.04.0: Released approximately before the breaking changes discussed for version NEXT:
  - Version NEXT YY.MM.DD.0: To be released after:
    - Chisel is bumped to 3.3
    - FIRRTL is bumped to X.XX
    - Command line interface updated for staged generator
    

> * Following versions released quarterly

Reviewer 2: If we're planning to do quarterly releases, is it necessary to include the day of the month in the version string? I have doubts to our ability to predict the exact day on which we will cut a release, and I'm not sure if that level of granularity warrants inclusion in the version string.

Author: I have zero intention of "predicting" the exact day we will cut a release. I think we should tag it when it happens
However I also think it's quite likely we will cut a release multiple times in a month (hoping that the features listed above happen before the end of April, as a very real example).

So maybe the bug in this section is really "following versions released quarterly" if the releases are going to really be feature-driven

Reviewer 2: > However at the current rate of change it's quite likely we will cut a release multiple times in a month (hoping that the features listed above happen before the end of April, as a very real example, though perhaps would stabilize after that?)

That was not the impression that I was getting from the document, since the only release cadence that was explicitly described was quarterly. Maybe the disconnect is that I was under the impression that SiFive internally would still be using SNAPSHOT/bleeding-edge releases, but that anyone (externally or internally) would have the option of using a more stable release created quarterly (and perhaps with bugfixes). There's always going to be a bit of a tradeoff between stability and new features, and that's why many large projects end up having different channels to cater to different needs.

Author: I'm coming around to agree it's pointless to put the day in there. 

Also, i think the concept of a "stable branch" is useful even if releases aren't getting cut from it per se.

Author: i removed the day from the propsoed date string

> * We attempt to deprecate APIs gracefully over at least 2 quarterly releases

Author: ```suggestion
* We attempt to deprecate APIs gracefully over at least 2 quarterly releases, but do not guarantee this.
```

Rocket Chip Versioning Strategy ** All Strawman to be discussed**
===============================

Version Naming
--------------

* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).
* The format is YYYY.0M.0D.MINOR
* Example:
  * 2020.04.04.0 is the release which occurred on April 4, 2020.
  * 2020.04.04.1 is a release that could occur at any time, but introduces only MINOR changes over the 2020.04.04.0 release
  * 2020.06.01.0 is a release that occurs on June 1, 2020, with MAJOR (API breaking) changes over the 2020.04.04.X release.
* MINOR changes include:
  * API Additions
  * Bug Fixes (which can result in changes in the emitted hardware considered bugs)
* MINOR changes do NOT include:
  * Backwards incompatible changes in the API (see below)

> 

Author: ```suggestion

If you would like to fork rocket-chip and release your own derivative versions, we suggest:
  * If you fork and cut a specific release off an existing commit off this repo:
    * Name it as <release>-SNAPSHOT-<git commit hash>
    * For example if you want to make a specific release off commit hash `1234beef` which is the next commit after the 2020.04.04.2 release, call it `2020.04.04.2-SNAPSHOT-1234beef`
  * If you fork and modify a release branch and want to make a release:
    * Name it as <release>-<identifier>-<arbitrary number with meaning decided by the entity doing the release>
  * For example, if GitHub user `@mwachs5` did this: 2020.04.04.2-mwachs5-0
  * For example, a project like SiFive Freedom did this: 2020.04.04.2-sifivefreedom-0
  * This document imposes no requirements/restrictions on the changes made in such forks, those can be managed by the projects/users as they see fit.

```

Author: I moved this further down

Rocket Chip Versioning Strategy ** All Strawman to be discussed**
===============================

Version Naming
--------------

* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).
* The format is YYYY.0M.0D.MINOR
* Example:
  * 2020.04.04.0 is the release which occurred on April 4, 2020.
  * 2020.04.04.1 is a release that could occur at any time, but introduces only MINOR changes over the 2020.04.04.0 release
  * 2020.06.01.0 is a release that occurs on June 1, 2020, with MAJOR (API breaking) changes over the 2020.04.04.X release.
* MINOR changes include:
  * API Additions
  * Bug Fixes (which can result in changes in the emitted hardware considered bugs)
* MINOR changes do NOT include:

>   * Changes in top-level I/Os for < level of hierarchy >

Author: ```suggestion
  * Changes in top-level I/Os for < level of hierarchy >
```

deleting this because it's captured (or not) in the API below

  * Backwards incompatible changes in the API (see below)

API
---

The Rocket-Chip API is: 
  * All public Scala methods and classes
  * All public Scala methods and classes of included Scala libraries (hardfloat, CDE, etc)
  * The command line used to invoke the generator
  * All Verilog interfaces (I/Os and parameterization) for verilog files
  * C code...?
  * The version(s) of Chisel, FIRRTL, rocket-tools pointed to by rocket-chip.
  * Any code that generates IO(s) of any Module
  * The cmd line interface to scripts like vlsi-mem-gen
  * `public val`s within `LazyModule`s which translate to diplomatic nodes (?? -- this could go either way)

The following is NOT part of the API:
  * `public val`s within `Module`s which translate to Wires, Regs, module instances

>     * (We note that this may cause issues for people using Cross-Module references)

Author: I'm thinking i should remove this relaxation so that we can use more automated tooling to find changes

Reviewer 4: What do you mean here?

Author: I *think* i am agreeing with your larger point below... people use the Scala API to do cross-module references, so we should NOT allow changes to `public vals` within Modules. This also lets us use normal binary compatability testing frameworks to detect these sort of changes.

None of this helps the person doing a verilog CMR, where things like annotations applied, .suggestName(s) applied, configuration settings might change, but this seems more straightforward to implement and audit to move this bullet up to the "part of the API" section above.

  * Internals of scripts like vlsi-mem-gen, the contents of such scripts
  * Internals of any verilog black boxes
  * C Code ...?
      
Roadmap
-------

* Upcoming Versions:

  - Version ~2020.04.04.0: Released approximately before the breaking changes discussed for version NEXT:

>   - Version NEXT YY.MM.DD.0: To be released after:

Author: ```suggestion
  - Version NEXT YY.0M.0D.0: To be released after:
```

The following is NOT part of the API:
  * `public val`s within `Module`s which translate to Wires, Regs, module instances
    * (We note that this may cause issues for people using Cross-Module references)
  * Internals of scripts like vlsi-mem-gen, the contents of such scripts
  * Internals of any verilog black boxes
  * C Code ...?
      
Roadmap
-------

* Upcoming Versions:

  - Version ~2020.04.04.0: Released approximately before the breaking changes discussed for version NEXT:
  - Version NEXT YYYY.0M.0D.0: To be released after:
    - Chisel is bumped to 3.3
    - FIRRTL is bumped to X.XX
    - Command line interface updated for staged generator
    
* Following versions released quarterly

* We attempt to deprecate APIs gracefully over at least 2 quarterly releases, but do not guarantee this.

Version Branches
-----------------

Development occurs for the next YYYY.0M.0D.0 release on the `master` branch.
When the YYYY.0M.0D.0 is made, a protected branch is created with the name: `YYYY.0M.0D.X`.
The MINOR releases will be made as tagged commits on `YYYY.0M.0D.X branch`.

> 

Author: ```suggestion
Naming Your Own Releases
---------------------------

If you would like to fork rocket-chip and release your own derivative versions, we suggest:
  * If you fork and cut a specific release off an existing commit off this repo:
    * Name it as <release>-SNAPSHOT-<git commit hash>
    * For example if you want to make a specific release off commit hash `1234beef` which is a commit after the 2020.04.04.2 release, call it `2020.04.04.2-SNAPSHOT-1234beef`
  * If you fork and modify a release branch and want to make a release:
    * Name it as <release>-<identifier>-<arbitrary number with meaning decided by the entity doing the release>
  * For example, if GitHub user `@mwachs5` did this: `2020.04.04.2-mwachs5-0`
 * This document suggests no requirements/restrictions on the changes made in such forks

Rocket Chip Versioning Strategy ** All Strawman to be discussed**
===============================

Version Naming
--------------

* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).

> * The format is YYYY.0M.0D.MINOR

Author: ```suggestion
* The format is YYYY.0M.MINOR
```

> * Example:
>   * 2020.04.04.0 is the release which occurred on April 4, 2020.
>   * 2020.04.04.1 is a release that could occur at any time, but introduces only MINOR changes over the 2020.04.04.0 release
>   * 2020.06.01.0 is a release that occurs on June 1, 2020, with MAJOR (API breaking) changes over the 2020.04.04.X release.

Author: ```suggestion
* Example:
  * 2020.04.0 is the release which occurred in April, 2020.
  * 2020.04.1 is a release that could occur at any time, but introduces only MINOR changes over the 2020.04.0 release
  * 2020.06.0 is a release that occurs in June 2020, with MAJOR (API breaking) changes over the 2020.04.X release(s).
```

* MINOR changes include:
  * API Additions
  * Bug Fixes (which can result in changes in the emitted hardware considered bugs)
* MINOR changes do NOT include:
  * Backwards incompatible changes in the API (see below)

API
---

The Rocket-Chip API is: 
  * All public Scala methods and classes
  * All public Scala methods and classes of included Scala libraries (hardfloat, CDE, etc)
  * The command line used to invoke the generator

>   * All Verilog interfaces (I/Os and parameterization) for verilog files

Author: ```suggestion
  * All Verilog interfaces (I/Os and parameterization) for verilog resources
```

  * C code...?
  * The version(s) of Chisel, FIRRTL, rocket-tools pointed to by rocket-chip.
  * Any code that generates IO(s) of any Module
  * The cmd line interface to scripts like vlsi-mem-gen
  * `public val`s within `LazyModule`s which translate to diplomatic nodes (?? -- this could go either way)

The following is NOT part of the API:
  * `public val`s within `Module`s which translate to Wires, Regs, module instances
    * (We note that this may cause issues for people using Cross-Module references)
  * Internals of scripts like vlsi-mem-gen, the contents of such scripts
  * Internals of any verilog black boxes

>   * C Code ...?

Author: ```suggestion
  * C Code ...?
  * Makefiles and Makefile fragments
  
```

>       

Author: ```suggestion
API Dependencies on included libraries/dependencies:
------------------------------------------------------

* A new MINOR release may update `firrtl`, `chisel`, and `rocket-tools` to a newer minor release of those dependencies.
* A new MAJOR release may update `firrtl`, `chisel`, and `rocket-tools` to a newer major release of those dependencies.

* A new MINOR release may update `wit` dependencies listed in `wit-manifest.json` to newer minor release of those dependencies.
* A new MAJOR release may update `wit` dependencies listed in `wit-manifest.json` to newer major release of those dependencies.
```

Roadmap
-------

* Upcoming Versions:

>   - Version ~2020.04.04.0: Released approximately before the breaking changes discussed for version NEXT:

Author: ```suggestion
  - Version 2020.04.0: To be released just before the breaking changes discussed for version NEXT:
```

>   - Version NEXT YYYY.0M.0D.0: To be released after:

Author: ```suggestion
  - Version NEXT YYYY.0M.0: To be released after:
```

    - Chisel is bumped to 3.3
    - FIRRTL is bumped to X.XX
    - Command line interface updated for staged generator
    
* Following versions released quarterly

* We attempt to deprecate APIs gracefully over at least 2 quarterly releases, but do not guarantee this.

Version Branches
-----------------

> Development occurs for the next YYYY.0M.0D.0 release on the `master` branch.

Author: ```suggestion
Development occurs for the next major YYYY.0M.0 release on the `master` branch.
```

> When the YYYY.0M.0D.0 is made, a protected branch is created with the name: `YYYY.0M.0D.X`.

Author: ```suggestion
When the YYYY.0M.0 release is made, a protected (no force-pushing) branch is created with the name: `YYYY.0M.X`.
```

> The MINOR releases will be made as tagged commits on `YYYY.0M.0D.X branch`.

Author: ```suggestion
The first MAJOR and subsequent MINOR releases will be made as tagged commits on `YYYY.0M.X branch`.
```

Naming Your Own Releases
---------------------------

If you would like to fork rocket-chip and release your own derivative versions, we suggest:
  * If you fork and cut a specific release off an existing commit off this repo:
    * Name it as <release>-SNAPSHOT-<git commit hash>

>     * For example if you want to make a specific release off commit hash `1234beef` which is a commit after the 2020.04.04.2 release, call it `2020.04.04.2-SNAPSHOT-1234beef`

Author: ```suggestion
    * For example if you want to make a specific release off commit hash `1234beef` which is a commit after the 2020.04.2 release, call it `2020.04.2-SNAPSHOT-1234beef`
```

>   * If you fork and modify a release branch and want to make a release:
>     * Name it as <release>-<identifier>-<arbitrary number with meaning decided by the entity doing the release>
>   * For example, if GitHub user `@mwachs5` did this: `2020.04.04.2-mwachs5-0`
>  * This document suggests no requirements/restrictions on the changes made in such forks

Author: ```suggestion
  * If you fork and modify a release branch and want to make a release:
    * Name it as <release>-<identifier>-<arbitrary number with meaning decided by the entity doing the release>
    * For example, if GitHub user `@mwachs5` did this: `2020.04.2-mwachs5-0`
 * This document suggests no requirements/restrictions on the changes made in such forks
```
